spec-examples/src/blake2/blake2b.rs

```diff
@@ -49,51 +73,75 @@ fn mix(v: DoubleState, a: usize, b: usize, c: usize, d: usize, x: U64, y: U64) -
     result
 }
 
-// TODO: add test case where counter wraps
-fn inc_counter(t: Counter, x: u64) -> Counter {
-    let mut result: Counter = Counter([0; 2]);
-    result[0] = t[0] + x;
-    if result[0] < x {
-        result[1] = t[1] + 1u64;
+fn inc_counter<Word: PublicInteger>(t: Counter<Word>, x: Word) -> Counter<Word> {
+    let mut result = Counter::new();
+    let new_val: Word = t[0] + x;
+    result[0] = new_val;
+    if new_val.less_than(x) {
+        result[1] = t[1] + Word::ONE;
     }
     result
 }
 
-// TODO: move to library
-fn make_u64array(h: Buffer) -> DoubleState {
+fn make_array<Word: UnsignedSecretInteger>(h: &ByteSeq) -> DoubleState<Word> {
+    assert_eq!(h.len() / ((Word::NUM_BITS as usize) / 8), 16);
     let mut result = DoubleState::new();
     for i in 0..16 {
-        result[i] = U64::from(h[8 * i])
-            | U64::from(h[1 + 8 * i]) << 8
-            | U64::from(h[2 + 8 * i]) << 16
-            | U64::from(h[3 + 8 * i]) << 24
-            | U64::from(h[4 + 8 * i]) << 32
-            | U64::from(h[5 + 8 * i]) << 40
-            | U64::from(h[6 + 8 * i]) << 48
-            | U64::from(h[7 + 8 * i]) << 56;
+        let (_, h_chunk) = h.get_chunk(Word::NUM_BITS as usize / 8, i);
+        result[i] = Word::from_le_bytes(&h_chunk)
     }
     result
 }
 
-fn compress(h: State, m: Buffer, t: Counter, last_block: bool) -> State {
+pub trait HasIV<Word: UnsignedSecretInteger> {
+    fn iv() -> State<Word>;
+}
+
+impl HasIV<U64> for State<U64> {
+    fn iv() -> State<U64> { State::from_seq(&IVB) }
+}
+
+impl HasIV<U32> for State<U32> {
+    fn iv() -> State<U32> { State::from_seq(&IVS) }
+}
+
+#[derive(Clone, Copy)]
+pub enum BlakeVariant {
+    Blake2S,
+    Blake2B,
+}
+
+fn compress<Word: UnsignedSecretInteger> (
+    h: State<Word>,
+    m: &ByteSeq,
+    t: Counter<Word::PublicVersion>,
+    last_block: bool,
+    alg: BlakeVariant,
+) -> State<Word> where State<Word>: HasIV<Word> {
     let mut v = DoubleState::new();
 
     // Read u8 data to u64.
-    let m = make_u64array(m);
+    let m = make_array(m);
 
     // Prepare.
     v = v.update_sub(0, &h, 0, 8);
-    v = v.update_sub(8, &IV, 0, 8);
-    v[12] ^= U64(t[0]);
-    v[13] ^= U64(t[1]);
+    v = v.update_sub(8, &State::iv(), 0, 8);
```

Comment:
Couldn't we handle `iv()` similar to the `num_rounds` below to avoid the trait?

Reply:
No we can't because `num_rounds` has the same type (`usize`) regardless of the Blake2 variant. However, the IV changes type depending on the variant (`State<U32>` or `State<U64>`)